presto-native-execution/presto_cpp/main/LinuxMemoryChecker.cpp

```diff
@@ -0,0 +1,183 @@
+#include <folly/File.h>
+#include <folly/FileUtil.h>
+#include <folly/Format.h>
+#include <folly/Range.h>
+#include <folly/String.h>
+#include <folly/gen/Base.h>
+#include <folly/gen/File.h>
+#include <folly/gen/String.h>
+#include "presto_cpp/main/PeriodicMemoryChecker.h"
+#include "presto_cpp/main/common/Configs.h"
+
+#include <sys/stat.h>
+
+namespace facebook::presto {
+
+class LinuxMemoryChecker : public PeriodicMemoryChecker {
+ public:
+  explicit LinuxMemoryChecker(
+      const PeriodicMemoryChecker::Config& config,
+      int64_t mallocBytes = 0,
+      std::function<void()>&& periodicCb = nullptr,
+      std::function<bool(const std::string&)>&& heapDumpCb = nullptr)
+      : PeriodicMemoryChecker(config),
+        mallocBytes_(mallocBytes),
+        periodicCb_(std::move(periodicCb)),
+        heapDumpCb_(std::move(heapDumpCb)) {}
+
+  ~LinuxMemoryChecker() override {}
+
+  void setMallocBytes(int64_t mallocBytes) {
+    mallocBytes_ = mallocBytes;
+  }
+
+  static std::unique_ptr<PeriodicMemoryChecker> createLinuxMemoryChecker() {
+    PeriodicMemoryChecker::Config config;
+    auto* systemConfig = SystemConfig::instance();
+    config.systemMemPushbackEnabled = systemConfig->systemMemPushbackEnabled();
+    config.systemMemLimitBytes = systemConfig->systemMemLimitGb() << 30;
+    config.systemMemShrinkBytes = systemConfig->systemMemShrinkGb() << 30;
+    return std::make_unique<LinuxMemoryChecker>(config);
+  }
+
+ protected:
+  // Current memory calculation used is inactive_anon + active_anon
+  // Our first attempt was using memInfo memTotal - memAvailable.
+  // However memInfo is not containerized so we reserve this as a
+  // last resort.
+  //
+  // Next we tried to use what docker/kubernetes uses for their
+  // calculation. cgroup usage_in_bytes - total_inactive_files.
+  // However we found out that usage_in_bytes is a fuzz value
+  // and has a chance for the sync to occur after the shrink
+  // polling interval. This would result in double shrinks.
+  //
+  // Therefore we decided on values from the memory.stat file
+  // that are real time statistics. At first we tried to use
+  // the calculation suggested by the kernel team RSS+CACHE(+SWAP)
+  // However we noticed that this value was not closely related to the
+  // value in usage_in_bytes which is used to OOMKill. We then looked
+  // at all of the values in the stat file and decided that
+  // inactive_anon + active_anon moves closest to that of
+  // usage_in_bytes
+  //
+  // NOTE: We do not know if cgroup V2 memory.current is a fuzz
+  // value. It may be better than what we currently use. For
+  // consistency we will match cgroup V1 and change if
+  // necessary.
+  int64_t systemUsedMemoryBytes() override {
+    static std::string statFile;
+    size_t memAvailable = 0;
+    size_t memTotal = 0;
+    size_t inactiveAnon = 0;
+    size_t activeAnon = 0;
+    std::array<char, 50> buf;
+    static const boost::regex inactiveAnonRegex(
```

Comment:
These statics should be outside the function at the top according to the coding conventions IMO. Please look at other such uses.